### PR TITLE
HDDS-6379. Not deducting the STANDALONE pipelines when counting pipelines on each datanode to check the pipeline limit

### PR DESCRIPTION
**Pull request by dombizita**
## What changes were proposed in this pull request?

In this change I modified the currentPipelineCount() method. I found that when the method calculated the current pipeline count in a datanode it counted in the STANDALONE pipelines. We shouldn't count the single node pipelines in, only the RATIS/THREE ones. So I renamed the method to currentRatisThreePipelineCount() and changed it to count the datanodes non closed RATIS/THREE pipelines. With this I fixed the problem, when we reached the pipeline limit with STANDALONE/ONE pipelines so it didn't create a RATIS/THREE pipeline. 

With the modified method some tests started failing in the TestPipelinePlacementPolicy. It was because some tests are using the insertHeavyNodesIntoNodeManager method which were only inserting the datanodes and its pipelines into the nodeManager and not into the pipelineStateManager. This wasn't a problem before because when we didn't find a pipeline in the pipelineStateManager we continued with the next pipeline, we didn't deducted that pipeline as a non single node pipeline, so we simply calculated it as a good one. Because of this I changed the insertHeavyNodesIntoNodeManager method to also add the pipelines to the pipelineStateManager. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6379

## How was this patch tested?

Added unit tests to the modified method.

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java:105]: I think this could be simplified to:

```
p.getReplicationConfig().equals(new RatisReplicationConfig(THREE)) && !p.isClosed()
```

And you could create a `private static final RATIS_THREE_CONFIG = new RatisReplicationConfig(THREE)` and use the constant in the above check to avoid creating a new object each time.

Comment (adoroszlai) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java:105]: > And you could create a `private static final RATIS_THREE_CONFIG = new RatisReplicationConfig(THREE)` and use the constant in the above check to avoid creating a new object each time.

I would like to suggest creating this as `public` in `RatisReplicationConfig`.  This would benefit other parts of Ozone code as well.

Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java:105]: Yea, I was just looking at the code to see where else we could benefit from this. I think it would be a good idea to a constant in RatisReplicationConfig for ONE and THREE. We could have a separate Jira to move any existing calls to `new RatisReplicationConfig...` over to the constant in non-test code, which might save some allocations.

I wonder if we would be better with a static method in RatisReplicationConfig, perhaps `getInstance(Factor factor)` - right now every ContainerInfo object in SCM holds a unique ReplicationConfig object - if we change the code correctly we could reduce all those duplicated instances down to 1. The object only holds a single field, but it will still make a small difference to memory.

I will create a followup Jira for this purpose and then we could implement that part separately to avoid blocking this one.

Comment (adoroszlai) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java:105]: > will create a followup Jira

Sure, I wasn't suggesting to perform all that refactoring in this PR.

Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java:105]: I went ahead and did the refactor in #3165 - the diff is large but repetitive. If we can commit it, then we could do the small refactor here. Lets see if the CI goes green in #3165 to start with!

Comment (dombizita) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java:105]: Thanks for the comments @sodonnel @adoroszlai, it would be simpler that way. Thanks for the followup Jira, I'll wait until the #3165 is committed.

Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java:105]: #3165 is committed now, so please go ahead and make the changes here.

**Review by sodonnel — APPROVED** [review]
LGTM, pending green CI.

**Comment by fapifta** [comment]
+1 I think the patch is good and also incorporates the changes made by Stephen.

Pending CI.

**Pull request merged**
